### ISSUE TITLE
Create Material model

### ISSUE THREAD

**Issue by redorav** [redorav/corsair-engine#29]
The material model governs the way geometry passes work. Getting this right is going to prove tricky so here's a summary.

## **CONCEPTS**
· **Vertex Buffer**: A source of geometry information
· **Shader**: The full pipeline (i.e. vertex + pixel shader, etc)
· **Render Mesh**: A single draw unit. Typically composed of multiple vertex buffers 
· **Material**: A collection of shaders (prepass, gbuffer, etc)
· **RenderModel**: A mesh and a material together. The render model produces pipeline objects that are directly usable for rendering. They can be compiled up front, in parallel, etc.

## **CREATION**
· A **vertex buffer** is created by loading a mesh. In the future we'll have an import process, but for now an fbx/gltf loads a vertex buffer
· A **mesh** is created by concatenating one or more vertex buffers
· A **shader** is creating by calling CrShaderCompiler and loading the returned binary object. A shader hasn't been compiled until it's a pipeline
· A **material** is loaded by loading an fbx/gltf at the moment (in the future the import process would create a default material for it). We'll call the Material Compiler with a descriptor which creates all the necessary shaders for it
· A **render model** is the final product after all those processes above. It has almost all the information needed to render, except for instance-specific data such as transform, skinning, and engine-provided data (e.g. refraction buffer)

## **LOADING**
Loading is probably the hardest part to define as it is context-defined.

**Meshes**: These are easier because there are two sources of geometry: source and serialized. In the first case we load from an FBX and create the mesh once the geometry is processed, in the second the geometry has been stored in a native format and just needs to be deserialized.

**Materials**: Material shaders are more complicated because there is a compilation process involved, there are multiple shaders per material, and there are multiple ways to obtain these shaders. A material typically contains all the shaders necessary to render an object in any given context (e.g. prepass, gbuffer, transparency). It has one shader for each of those contexts and knows all the render state that is necessary (render target formats, depth state, etc). Materials are built through a material descriptor that kicks off a process to obtain the shaders for it. There are two ways to obtain shaders.

a) Source: If the shader binary isn't available anywhere, load the source file that contains the entry point, set up defines as necessary and call CrShaderCompiler. This in turn calls the platform-specific compiler and produces an output where the application is expecting it. After that the application loads the binary and passes it to the API. The shader manager can have a cache in temp and a runtime cache to accelerate the lookups. The key is the entire command line and it needs to compare the date.

b) Compiled: Load the binary objects. These binaries are ready to use by the API (bar some runtime patching e.g. vulkan bind points) These binaries live in the data and be part of some binary blob, etc. This would be a second step when importing and creating materials is more defined.

**Pipelines**: Each render model points to its meshes, its materials (that meshes use) and it creates pipelines for the combinations. These pipelines are created when the render model is instantiated or in a deferred manner, in any case they need to be ready before any rendering happens to avoid stalls. The pipeline manager already has a cache to avoid creating the same pipelines repeatedly, and has the offline cache as well.
